pkg/kv/store.go

```diff
@@ -0,0 +1,114 @@
+type Driver interface {
+	// Open access to the database store. Implementation should give access to the same storage based on the dsn.
+	// Implementation can return the same Storage instance based on dsn or new one as long as it provides access to
+	// the same storage.
+	Open(ctx context.Context, dsn string) (Store, error)
+}
+
+type Store interface {
+	// Get returns a value for the given key, or ErrNotFound if key doesn't exist
+	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Set stores the given value, overwriting an existing value if one exists
+	Set(ctx context.Context, key, value []byte) error
+
+	// SetIf returns an ErrPredicateFailed error if the key with valuePredicate passed
+	//  doesn't match the currently stored value. SetIf is a simple compare-and-swap operator:
+	//  valuePredicate is either the existing value, or nil for no previous key exists.
+	//  this is intentionally simplistic: we can model a better abstraction on top, keeping this interface simple for implementors
+	SetIf(ctx context.Context, key, value, valuePredicate []byte) error
+
+	// Delete will delete the key, no error in if key doesn't exist
+	Delete(ctx context.Context, key []byte) error
+
+	// Scan returns entries that can be read by key order, starting at or after the `start` position
+	Scan(ctx context.Context, start []byte) (Entries, error)
+
+	// Close access to the database store. After calling Close the instance is unusable.
+	Close()
+}
+
+// Entries used to enumerate over Scan results
+type Entries interface {
+	// Next should be called first before access Entry.
+	// it will process the next entry and return true if it was successful, and false when none or error.
+	Next() bool
+
+	// Entry current entry read after calling Next, set to nil in case of an error or no more entries.
+	Entry() *Entry
+
+	// Err set to last error by reading or parse the next entry.
+	Err() error
+
+	// Close should be called at the end of processing entries, required to release resources used to scan entries.
+	// After calling 'Close' the instance should not be used as the behaviour will not be defined.
+	Close()
+}
+
+// Entry holds a pair of key/value
+type Entry struct {
+	Key   []byte
+	Value []byte
+}
+
+func (e *Entry) String() string {
+	if e == nil {
+		return "Entry{nil}"
+	}
+	return fmt.Sprintf("Entry{%s, %s}", e.Key, e.Value)
+}
+
+// map drivers implementation
+var (
+	drivers   = make(map[string]Driver)
+	driversMu sync.RWMutex
+)
+
+func Register(name string, driver Driver) {
+	if name == "" {
+		panic("kv store register name is missing")
+	}
+	if driver == nil {
+		panic("kv store Register driver is nil")
+	}
+	driversMu.Lock()
+	defer driversMu.Unlock()
+	if _, found := drivers[name]; found {
+		panic("kv store Register driver already registered " + name)
+	}
+	drivers[name] = driver
+}
+
+func Open(ctx context.Context, name, dsn string) (Store, error) {
+	driversMu.RLock()
+	d, ok := drivers[name]
+	driversMu.RUnlock()
```

Comment:
```suggestion
	driversMu.Lock()
	d, ok := drivers[name]
	driversMu.Unlock()
```
and switch to a RW-mutex later, if it is actually any faster.  I worry that we do not spend enough time in the lock to want to bother will full RW mutex fairness.

Reply:
https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/database/sql/sql.go;l=64
Good for Go's database/sql - think it will be good for us. Also I don't predict performance issue as we keep reference to a Store and not calling Open all the time during the current flow of the code.

Reply:
Sure, just seriously not a fan of RWMutex unless absolutely indicated by tests.  I just ran this benchmark with varying numbers of concurrent readers and writers:
[mutex_test.go.gz](https://github.com/treeverse/lakeFS/files/8560366/mutex_test.go.gz).  You need to tune to ~10x more readers than writers to start winning on a plain map.  (And if it's just called at startup, performance doesn't matter and the simpler Mutex wins.)